src/crypto/nss.rs

```diff
@@ -87,123 +96,445 @@ pub(crate) fn serialize_key(_curve: ECDSACurve, key: &[u8]) -> Result<(ByteBuf,
-    let ec_key = EcKey::new(curve, my_priv_key, my_pub_key);
-    let private_key = PrivateKey::import(&ec_key)?;
-    encapsulate_helper(
-        key,
+    curve: ECDSACurve,
+    nss_public: PublicKey,
+) -> Result<COSEKey> {
+    let public_data = nss_public.key_data()?;
+    let (public_x, public_y) = serialize_key(curve, &public_data)?;
+    Ok(COSEKey {
         alg,
-        my_pub_key,
-        private_key._tests_only_dangerously_convert_to_ephemeral(),
-    )
+        key: COSEKeyType::EC2(COSEEC2Key {
+            curve,
+            x: public_x.to_vec(),
+            y: public_y.to_vec(),
+        }),
+    })
+}
+
+/// `peer_public`: The authenticator's public key.
+/// `client_private`: Our ephemeral private key.
+fn encapsulate_helper(peer_public: PublicKey, client_private: PrivateKey) -> Result<Vec<u8>> {
+    let ecdh_x_coord = unsafe {
+        PK11_PubDeriveWithKDF(
+            *client_private,
+            *peer_public,
+            PR_FALSE,
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+            CKM_ECDH1_DERIVE,
+            CKM_SHA512_HMAC, // unused
+            CKA_DERIVE,      // unused
+            0,
+            CKD_NULL,
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+        )
+        .into_result()?
+    };
+    let ecdh_x_coord_bytes = ecdh_x_coord.as_bytes()?;
+    let mut shared_secret = [0u8; SHA256_LENGTH as usize];
+    unsafe {
+        PK11_HashBuf(
+            SECOidTag::SEC_OID_SHA256,
+            shared_secret.as_mut_ptr(),
+            ecdh_x_coord_bytes.as_ptr(),
+            ecdh_x_coord_bytes.len() as i32,
+        )
+    };
+    Ok(shared_secret.to_vec())
 }
 
 /// Encrypts a plaintext to produce a ciphertext, which may be longer than the plaintext.
 /// The plaintext is restricted to being a multiple of the AES block size (16 bytes) in length.
-pub(crate) fn encrypt(
-    key: &[u8],
-    plain_text: &[u8], /*PlainText*/
-) -> Result<Vec<u8> /*CypherText*/> {
-    crypt_helper(key, plain_text, Operation::Encrypt)
+pub(crate) fn encrypt(key: &[u8], data: &[u8]) -> Result<Vec<u8>> {
+    nss_gk_api::init();
+
+    if key.len() != 32 {
+        return Err(BackendError::NSSError(
+            "Invalid AES-256 key length".to_string(),
+        ));
+    }
+
+    // The input must be a multiple of the AES block size, 16
+    if data.len() % (AES_BLOCK_SIZE as usize) != 0 {
+        return Err(BackendError::NSSError(
+            "Input to encrypt is too long".to_string(),
+        ));
+    }
+    let in_len = c_uint::try_from(data.len())?;
+
+    let slot = Slot::internal()?;
+
+    let sym_key = unsafe {
+        PK11_ImportSymKey(
+            *slot,
+            CKM_AES_CBC,
+            PK11Origin::PK11_OriginUnwrap,
+            CKA_ENCRYPT,
+            SECItemBorrowed::wrap(key).as_mut(),
+            ptr::null_mut(),
+        )
+        .into_result()?
+    };
+
+    let iv = [0u8; AES_BLOCK_SIZE as usize];
```

Comment:
The repeated `as usize` is a bit of a bummer - maybe `nss-gk-api` could define some helper constants? Not a big deal, though.

Reply:
Agreed. I opened https://github.com/mozilla/nss-gk-api/issues/2